coil-compose-core/src/commonMain/kotlin/coil3/compose/AsyncImagePainter.kt

```diff
@@ -159,14 +153,9 @@ class AsyncImagePainter internal constructor(
         replay = 1,
         onBufferOverflow = DROP_OLDEST,
     )
-    private val restartSignal = MutableSharedFlow<Unit>(
-        replay = 1,
-        onBufferOverflow = DROP_OLDEST,
-    ).apply { tryEmit(Unit) }
-
     private var painter: Painter? by mutableStateOf(null)
-    private var alpha: Float by mutableFloatStateOf(DefaultAlpha)
-    private var colorFilter: ColorFilter? by mutableStateOf(null)
+    private var alpha: Float = DefaultAlpha
+    private var colorFilter: ColorFilter? = null
```

Comment:
I think we might have to notify the composition if `alpha`/`colorFilter` changes so that `DrawScope.onDraw` is reinvoked. Or is that handled elsewhere?

Reply:
they are not even changed during the modifier update, those two are always updated during drawing right before drawing the painter content itself: https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:compose/ui/ui-graphics/src/commonMain/kotlin/androidx/compose/ui/graphics/painter/Painter.kt;l=173;drc=dcaa116fbfda77e64a319e1668056ce3b032469f. because of that we don't need to invalidate on change.
and also, by using mutables states here we might even cause extra invalidation on every change as we update the state, and then read straight away. then when the state change applied we invalidate the drawing one more time as we had a read for that